### PR TITLE
block: qcow: Add RAW backing file support for QCOW2 images

### PR DESCRIPTION
**Pull request by weltling**
This patch series implements support for RAW backing files in QCOW2 images, enabling QCOW2 overlays to use RAW images as their backing store.

- Add RAW format backing file support alongside existing QCOW2 backing support
- Implement QCOW2 v3 header extension parsing to read backing file format specifications
- For v2 and below, add format auto-detection when backing format is not explicitly specified
- Add corresponding functional and performance tests along with some unit test fixes

Performance tests comparing QCOW2 vs RAW backing files show RAW backing provides 43% better sequential read performance (874 vs. 1254 MiB/s) and 5% better random read performance (1625 vs. 1704 MiB/s). Future optimization opportunities remain.

### CODE REVIEW

**Comment by liuw** [comment]
We put in the `qemu-img` check in the test suite, but does it check the backing file's consistency? If not, we should find a way to check that, too.

**Comment by weltling** [comment]
> We put in the `qemu-img` check in the test suite, but does it check the backing file's consistency? If not, we should find a way to check that, too.

Yep, that's a good point. While the backing file itself is readonly, of course one could add some validation in the tests. Fro the runtime, I've noted this, too, will check if some fast method is possible without impacting the perf much, or maybe as an extra option.

Thanks

**Comment by weltling** [comment]
I've now added now the backing file checksum validaiton to the tests. For qcow additionally also `qemu-img check ...`, too.

Thanks

File: block/src/qcow/mod.rs [hunk 1]
```diff
@@ -171,6 +173,10 @@ const COMPATIBLE_FEATURES_LAZY_REFCOUNTS: u64 = 1;
 const COMPRESSION_TYPE_ZLIB: u64 = 0; // zlib/deflate <https://www.ietf.org/rfc/rfc1951.txt>
 const COMPRESSION_TYPE_ZSTD: u64 = 1; // zstd <http://github.com/facebook/zstd>
 
+// Header extension types
+const HEADER_EXT_END: u32 = 0x00000000;
+const HEADER_EXT_BACKING_FORMAT: u32 = 0xe2792aca;
```
Comment (russell-islam) [inline block/src/qcow/mod.rs:208]: Add a comment on what format.

File: block/src/qcow/mod.rs [hunk 2]
```diff
@@ -244,9 +250,49 @@ pub struct QcowHeader {
 
     // Post-header entries
     pub backing_file_path: Option<String>,
+    pub backing_file_format: Option<String>,
 }
 
 impl QcowHeader {
+    fn read_header_extensions(f: &mut RawFile, header: &mut QcowHeader) -> Result<()> {
```
Comment (russell-islam) [inline block/src/qcow/mod.rs:286]: Why do you need a mut reference where you just read the contents? Is it possible to adjust offset from the caller if that's the need?

File: block/src/qcow/mod.rs [hunk 3]
```diff
@@ -479,6 +544,98 @@ fn max_refcount_clusters(refcount_order: u32, cluster_size: u32, num_clusters: u
...
+        if max_nesting_depth == 0 {
+            return Err(Error::MaxNestingDepthExceeded);
+        }
+
+        let backing_raw_file = OpenOptions::new()
+            .read(true)
+            .open(path)
+            .map_err(Error::BackingFileIo)?;
+
+        let mut raw_file = RawFile::new(backing_raw_file, direct_io);
```
Comment (russell-islam) [inline block/src/qcow/mod.rs:622]: Why is direct IO? Please explain.

File: block/src/qcow/mod.rs [hunk 4]
```diff
@@ -1963,6 +2102,14 @@ pub fn detect_image_type(file: &mut RawFile) -> Result<ImageType> {
     Ok(image_type)
 }
 
+/// Get the image type as a string ("raw" or "qcow2")
+fn get_image_type_str(file: &mut RawFile) -> Result<&'static str> {
+    match detect_image_type(file)? {
+        ImageType::Qcow2 => Ok("qcow2"),
+        ImageType::Raw => Ok("raw"),
+    }
```
Comment (russell-islam) [inline block/src/qcow/mod.rs]: no fall-back error?

File: block/src/qcow/mod.rs (hunk 2 above)
Comment (weltling) [reply block/src/qcow/mod.rs:286]: Yep, both are needed. The header is the subject of the modification. With the offset - nope, it needs to seek through multiple places in the function. 

Thanks

File: block/src/qcow/mod.rs (hunk 3 above)
Comment (weltling) [reply block/src/qcow/mod.rs:622]: This comes from the pre-existing code. I'm also not sure if buffered I/O would make a diff, but IMO the first implementation where i'd rather use the inherited code as is. I'd go for optimizations later now that some perf infrastructure is in place.

thanks

File: block/src/qcow/mod.rs (hunk 1 above)
Comment (weltling) [reply block/src/qcow/mod.rs:208]: Fixed, thanks.

File: block/src/qcow/mod.rs (hunk 4 above)
Comment (weltling) [reply block/src/qcow/mod.rs]: I made the format test stricter, now it'll explicitly handle this through `ImageType::Unknown`.

Thanks

Comment (weltling) [reply block/src/qcow/mod.rs]: FYI I reverted this approach for now. It would have been a nice quick refactoring, but there are seem to be more tricky cases outside the qcow module. The current logic is to treat everthing not explicitly `qcow` as `raw`. I noted this case for a separate refactoring, keeping the current logic for this initial implementation .

Thanks

File: block/src/qcow/mod.rs [hunk 5]
```diff
@@ -479,6 +545,98 @@ fn max_refcount_clusters(refcount_order: u32, cluster_size: u32, num_clusters: u
...
+    fn clone_box(&self) -> Box<dyn BackingFileOps> {
+        Box::new(self.clone())
+    }
+}
+
+impl BackingFileOps for RawFile {
+    fn clone_box(&self) -> Box<dyn BackingFileOps> {
+        Box::new(self.clone())
+    }
+}
```
Comment (russell-islam) [inline block/src/qcow/mod.rs:595]: Can we add these functions to the trait and just impl BackingFileOps for QcowFile|RawFile {}

Comment (weltling) [reply block/src/qcow/mod.rs:595]: Yea, I was puzzling about this part as well and nope, unfortunately a default implementation is not possible in this case. Here is why. The `clone_box()` is needed, as `QcowFile` derives `Clone` and `BackingFile` has been made alike. `BackingFileOps` is used as a trait object `Box<dyn BackingFileOps>` and it would need to add `Clone` to the trait bounds. However, the `Clone` trait is not object safe as it requires `fn clone(&self) -> Self` where `Self` is a value. Rust refuses to compile it as a common method, as the size of `Self` is unknown at the compile time.

If we tried:

```rust
trait BackingFileOps: Send + Seek + Read + Clone {
    fn clone_box(&self) -> Box<dyn BackingFileOps> {
        Box::new(self.clone())
    }
}
```
We'd get:

```
error[E0038]: the trait `BackingFileOps` cannot be made into an object
   --> block/src/qcow/mod.rs:XXX:XX
    |
    | trait BackingFileOps: Send + Seek + Read + Clone {
    |                                             ^^^^^ `BackingFileOps` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically
   --> block/src/qcow/mod.rs:XXX:XX
    |
    = note: the trait cannot require that `Self : Sized`
```

Thus, an object specific implementation is a compilation time requirement. Alternatively one could switch to using enum instead of trait objects would avoid this issue but would introduce match overhead on every IO operation, which isn't desirable for a hot path.

Thanks

File: cloud-hypervisor/tests/integration.rs [hunk 6]
```diff
@@ -3517,35 +3526,119 @@ mod common_parallel {
...
         } else {
             // An image name is provided
             let mut workload_path = dirs::home_dir().unwrap();
             workload_path.push("workloads");
             workload_path.as_path().join(path_or_image_name.as_ref())
+        }
+    }
+
+    fn compute_file_checksum(path: &std::path::Path) -> u32 {
+        use std::io::Read;
```
Comment (rbradford) [inline cloud-hypervisor/tests/integration.rs]: nit: I think we normally put use statements outside the method.

File: cloud-hypervisor/tests/integration.rs [hunk 7]
```diff
@@ -3517,35 +3526,119 @@ mod common_parallel {
...
+    ///
+    /// For qcow2 images with backing files, also verifies the backing file
+    /// integrity and checks that the backing file hasn't been modified
+    /// during the test.
+    fn disk_check_consistency(
+        path_or_image_name: impl AsRef<std::path::Path>,
+        initial_backing_checksum: Option<(std::path::PathBuf, String, u32)>,
+    ) {
+        let path = resolve_disk_path(path_or_image_name);
+
```
Comment (rbradford) [inline cloud-hypervisor/tests/integration.rs]: nit: This whitespace line feels unnecessary.

File: block/src/qcow/mod.rs [hunk 8]
```diff
@@ -244,6 +246,7 @@ pub struct QcowHeader {
 
     // Post-header entries
     pub backing_file_path: Option<String>,
+    pub backing_file_format: Option<String>,
```
Comment (rbradford) [inline block/src/qcow/mod.rs]: What's that rationale for this being a `String` and not an enum type?

File: performance-metrics/src/main.rs [hunk 9]
```diff
@@ -319,7 +319,7 @@ mod adjuster {
     }
 }
 
-const TEST_LIST: [PerformanceTest; 32] = [
+const TEST_LIST: [PerformanceTest; 34] = [
```
Comment (rbradford) [inline performance-metrics/src/main.rs:322]: @likebreath Needs to review this file.

File: block/src/qcow/mod.rs [hunk 10]
```diff
@@ -250,6 +255,45 @@ pub struct QcowHeader {
 }
 
 impl QcowHeader {
+    fn read_header_extensions(f: &mut RawFile, header: &mut QcowHeader) -> Result<()> {
```
Comment (rbradford) [inline block/src/qcow/mod.rs:286]: Ideally this function would be unit tested.

File: cloud-hypervisor/tests/integration.rs [hunk 11]
```diff
@@ -3517,35 +3526,119 @@ mod common_parallel {
...
             path_or_image_name.as_ref().to_path_buf()
         } else {
             // An image name is provided
             let mut workload_path = dirs::home_dir().unwrap();
             workload_path.push("workloads");
             workload_path.as_path().join(path_or_image_name.as_ref())
+        }
+    }
+
+    fn compute_file_checksum(path: &std::path::Path) -> u32 {
```
Comment (rbradford) [inline cloud-hypervisor/tests/integration.rs]: This should probably take `& std::io::Read` and then be unit tested.

File: cloud-hypervisor/tests/integration.rs (hunk 7 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: Fixed, thanks.

File: cloud-hypervisor/tests/integration.rs (hunk 6 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: Fixed, thanks! I've overseen it's already imported through a grouped import at the top.

File: cloud-hypervisor/tests/integration.rs (hunk 11 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: Done. Added unit tests for deterministic hashing and edge cases. Since this is test infrastructure, the unit tests provide extra confidence. Thanks

File: block/src/qcow/mod.rs (hunk 8 above)
Comment (weltling) [reply block/src/qcow/mod.rs]: Good point! While QCOW2 stores it as a string in the header extension, it's effectively a fixed set of formats (raw/qcow2). I've updated it to use `ImageType` enum with `FromStr`/`Display` traits. Thanks!

File: block/src/qcow/mod.rs (hunk 10 above)
Comment (weltling) [reply block/src/qcow/mod.rs:286]: ACK. I've added some unit tests specifically for this function, which can be extended once more extensions are parsed. Thanks

**Comment by weltling** [comment]
> @weltling It's really cool to see improvements to the QCOW2 code - you're becoming our goto expert on that! Since this is pretty isolated code there is good scope for unit testing (vs other parts of the codebase) so i'd encourage you to include more of that.

Thanks! Yep, there are definitely more tests to come and I've already some improvement ideas based on the previous PR conversations. For QCOW2 itself, as the perf validation baseline is being established, I'd anticipate more to achieve especially for the RAW backing file. But also, with #7560 there is yet another wave of improvements to come for the whole `./block` module.

File: block/src/qcow/mod.rs [hunk 12]
```diff
@@ -244,6 +268,7 @@ pub struct QcowHeader {
 
     // Post-header entries
     pub backing_file_path: Option<String>,
+    pub backing_file_format: Option<ImageType>,
```
Comment (phip1611) [inline block/src/qcow/mod.rs]: Would it make sense to introduce a new type `BackingFileConfig` or so type that bundles these properties? Otherwise, one has to  check in code for situations where `backing_file_format` is Some and `backing_file_path` is None (or vice versa) which to my understanding is invalid anyway.

File: cloud-hypervisor/tests/integration.rs [hunk 13]
```diff
@@ -3517,35 +3526,170 @@ mod common_parallel {
...
             // An image name is provided
             let mut workload_path = dirs::home_dir().unwrap();
             workload_path.push("workloads");
             workload_path.as_path().join(path_or_image_name.as_ref())
+        }
+    }
+
+    fn compute_file_checksum(reader: &mut dyn std::io::Read, size: u64) -> u32 {
+        // Read first 16MB or entire data if smaller
+        let read_size = std::cmp::min(size, 16 * 1024 * 1024) as usize;
```
Comment (phip1611) [inline cloud-hypervisor/tests/integration.rs]: nit, personal style preference, feel free to ignore: `std::cmp::min` -> `cmp::min` + `use std::cmp`

File: cloud-hypervisor/tests/integration.rs [hunk 14]
```diff
@@ -3517,35 +3526,170 @@ mod common_parallel {
...
+        let mut hash: u32 = 5381;
+        for byte in buffer.iter() {
+            hash = hash.wrapping_mul(33).wrapping_add(*byte as u32);
+        }
+        hash
+    }
+
+    #[test]
+    fn test_compute_file_checksum_empty() {
+        let mut reader = std::io::Cursor::new(vec![]);
```
Comment (phip1611) [inline cloud-hypervisor/tests/integration.rs]: nit, personal style preference, feel free to ignore: `std::io::Cursor` -> `io::Cursor` + `use std::io`

File: cloud-hypervisor/tests/integration.rs [hunk 15]
```diff
@@ -3517,35 +3526,170 @@ mod common_parallel {
...
+                .join(backing_file)
         };
 
-        let output = std::process::Command::new("qemu-img")
-            .args(["check", path.to_str().unwrap()])
-            .output()
-            .expect("should spawn and run command successfully");
+        let backing_info = get_image_info(&backing_path)?;
+        let backing_format = backing_info["format"].as_str()?.to_string();
+        let mut file = std::fs::File::open(&backing_path).ok()?;
```
Comment (phip1611) [inline cloud-hypervisor/tests/integration.rs]: nit: `std::fs::File::open` -> `File::open`

File: cloud-hypervisor/tests/integration.rs [hunk 16]
```diff
@@ -3517,35 +3526,170 @@ mod common_parallel {
...
+        let output = run_qemu_img(&path, &["check"]);
 
         assert!(
             output.status.success(),
             "qemu-img check failed: {}",
             String::from_utf8_lossy(&output.stderr)
         );
+
+        if let Some((backing_path, format, initial_checksum)) = initial_backing_checksum {
+            if format != "raw" {
```
Comment (phip1611) [inline cloud-hypervisor/tests/integration.rs]: I am somewhat concerned that this relies on a "magic value" rather than a well-defined type, constant, or enum variant. However, I am not that familiar with this part of the codebase to judge whether this can be improved easily as a drive-by change.

**Review by phip1611 — CHANGES_REQUESTED** [review]
Generally LGTM, thanks for working on this! I left one remark and a few nits.

File: cloud-hypervisor/tests/integration.rs (hunk 16 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: The format string comes from qemu-img output, that's how it is stored in the image header. But this brought me to another idea - I've added `block` to the dev dependencies and switched to comparing using `block::qcow::ImageType`. That's of course much cleaner and indirectly tests the block crate. Otherwise, this condition check is uncritical as it's only about deciding in the test code whether `qemu-img check ..` is to be run. Thanks!

File: cloud-hypervisor/tests/integration.rs (hunk 15 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: Fixed, thanks.

File: cloud-hypervisor/tests/integration.rs (hunk 14 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: Fixed, thanks.

File: cloud-hypervisor/tests/integration.rs (hunk 13 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: Fixed, thanks.

File: block/src/qcow/mod.rs (hunk 12 above)
Comment (weltling) [reply block/src/qcow/mod.rs]: Yeah, there can be no backing file path without the format and vice versa. Implemented as per your suggestion. Thanks!

File: cloud-hypervisor/tests/integration.rs (hunk 16 above)
Comment (phip1611) [reply cloud-hypervisor/tests/integration.rs]: love it, awesome!

**Comment by phip1611** [comment]
@weltling Since the CI is still failing, here is a small hint: we have invested significant effort into improving developer productivity (#7489). You can just run `cargo clippy` locally, which should now work (this was not the case prior to v50).

**Comment by weltling** [comment]
> @weltling Since the CI is still failing, here is a small hint: we have invested significant effort into improving developer productivity (#7489). You can just run `cargo clippy` locally, which should now work (this was not the case prior to v50).

Thanks for pointing that out, @phip1611. I indeed looked up the clippy cmd from a failing job. And previously the other one about the bisectability. Got a little bit complicated this round, as many changes were to re-integrate into the history while keeping the history itself untouched, so caught some issues due to the cmoplexity. But should be all fixed now and good to know for the future more things work locally now.

Thanks!

**Comment by phip1611** [comment]
> > @weltling Since the CI is still failing, here is a small hint: we have invested significant effort into improving developer productivity (#7489). You can just run `cargo clippy` locally, which should now work (this was not the case prior to v50).
> 
> Thanks for pointing that out, @phip1611. I indeed looked up the clippy cmd from a failing job. And previously the other one about the bisectability. Got a little bit complicated this round, as many changes were to re-integrate into the history while keeping the history itself untouched, so caught some issues due to the cmoplexity. But should be all fixed now and good to know for the future more things work locally now.
> 
> Thanks!

I recently learned about `git rebase -i HEAD~10 --exec "cargo check && cargo +nightly fmt && cargo clippy"`. Game changer! It is so easy to style check every single commit and easily amend the corresponding changes

(there might be minor typos in the command as I'm on the phone)

File: performance-metrics/src/main.rs (hunk 9 above)
Comment (weltling) [reply performance-metrics/src/main.rs:322]: Hey @likebreath, would you have time to check this part please? Thanks

**Comment by weltling** [comment]
> > > @weltling Since the CI is still failing, here is a small hint: we have invested significant effort into improving developer productivity (#7489). You can just run `cargo clippy` locally, which should now work (this was not the case prior to v50).
> > 
> > 
> > Thanks for pointing that out, @phip1611. I indeed looked up the clippy cmd from a failing job. And previously the other one about the bisectability. Got a little bit complicated this round, as many changes were to re-integrate into the history while keeping the history itself untouched, so caught some issues due to the cmoplexity. But should be all fixed now and good to know for the future more things work locally now.
> > Thanks!
> 
> I recently learned about `git rebase -i HEAD~10 --exec "cargo check && cargo +nightly fmt && cargo clippy"`. Game changer! It is so easy to style check every single commit and easily amend the corresponding changes
> 
> (there might be minor typos in the command as I'm on the phone)

So far very useful, thanks for the hint. The only pity is it doesn't work in the config `rebase.exec`. Thanks

File: block/src/qcow/mod.rs [hunk 17]
```diff
@@ -125,18 +128,45 @@ pub enum Error {
...
 #[derive(Clone, Debug)]
 pub enum CompressionType {
     Zlib,
     Zstd,
 }
 
+#[derive(Debug, Clone)]
+pub struct BackingFileConfig {
+    pub path: String,
+    pub format: Option<ImageType>,
```
Comment (phip1611) [inline block/src/qcow/mod.rs:168]: nit: 

```suggestion
    // If this is None, we will autodetect it.
    pub format: Option<ImageType>,
```

feel free to ignore

File: cloud-hypervisor/tests/integration.rs [hunk 18]
```diff
@@ -2527,8 +2527,9 @@ EOF
 }
 
 mod common_parallel {
-    use std::fs::OpenOptions;
-    use std::io::SeekFrom;
+    use std::cmp;
```
Comment (phip1611) [inline cloud-hypervisor/tests/integration.rs:2530]: not about the line, but the commit message. looks kinda broken?!

Image

File: cloud-hypervisor/tests/integration.rs [hunk 19]
```diff
@@ -3517,35 +3527,170 @@ mod common_parallel {
...
+        assert_eq!(position, 16 * 1024 * 1024);
+    }
+
+    fn compute_backing_checksum(
+        path_or_image_name: impl AsRef<std::path::Path>,
+    ) -> Option<(std::path::PathBuf, String, u32)> {
+        let path = resolve_disk_path(path_or_image_name);
+
+        let info = get_image_info(&path)?;
+        if info["format"].as_str()? != "qcow2" {
```
Comment (phip1611) [inline cloud-hypervisor/tests/integration.rs]: nit: might be a usecase for `detect_image_type()`?

File: cloud-hypervisor/tests/integration.rs (hunk 18 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs:2530]: Yep. Fixed that, thanks.

File: cloud-hypervisor/tests/integration.rs (hunk 19 above)
Comment (weltling) [reply cloud-hypervisor/tests/integration.rs]: Yep, it's definitely a use case. The suggestion is implemented. Thanks
